### ISSUE TITLE
In rare cases jedi calls pandas.DataFrame.T which can be very slow

### ISSUE THREAD

**Issue by arieleiz** [davidhalter/jedi#1933]
Once in a while we see ipykernel getting stuck, and the stack trace shows it is transposing the pandas dataframe from within `jedi.inference.compiled.create_from_name`. This can be very long (>10 minutes) for large data structures. will appreciate any ideas. thanks!

```
Thread 2348866 (active+gil): "MainThread"
    astype (pandas/core/arrays/datetimelike.py:415)
    astype (pandas/core/arrays/datetimes.py:666)
    get_values (pandas/core/internals/blocks.py:1451)
    _interleave (pandas/core/internals/managers.py:1636)
    as_array (pandas/core/internals/managers.py:1589)
    values (pandas/core/frame.py:10877)
    transpose (pandas/core/frame.py:3397)
    T (pandas/core/frame.py:3406)
    getattr_paths (jedi/inference/compiled/access.py:366)
    get_compiled_method_return (jedi/inference/compiled/subprocess/functions.py:27)
    _cached_results (jedi/inference/compiled/subprocess/__init__.py:391)
    wrapper (jedi/cache.py:112)
    _workaround (jedi/inference/compiled/subprocess/__init__.py:387)
    create_from_name (jedi/inference/compiled/value.py:574)
    infer_compiled_value (jedi/inference/compiled/value.py:352)
    wrapper (jedi/cache.py:112)
    api_type (jedi/inference/compiled/value.py:345)
    __getattr__ (jedi/inference/names.py:629)
    type (jedi/api/classes.py:193)
    type (jedi/api/classes.py:725)
    _completions (IPython/core/completer.py:1965)
    completions (IPython/core/completer.py:1909)
    rectify_completions (IPython/core/completer.py:501)
    _experimental_do_complete (ipykernel/ipkernel.py:475)
    do_complete (ipykernel/ipkernel.py:444)
    complete_request (ipykernel/kernelbase.py:769)
    dispatch_shell (ipykernel/kernelbase.py:398)
    process_one (ipykernel/kernelbase.py:491)
    dispatch_queue (ipykernel/kernelbase.py:502)
    _run (asyncio/events.py:81)
    _run_once (asyncio/base_events.py:1859)
    run_forever (asyncio/base_events.py:570)
    start (tornado/platform/asyncio.py:215)
    start (ipykernel/kernelapp.py:707)
    launch_instance (traitlets/config/application.py:992)
    <module> (ipykernel_launcher.py:17)
    _run_code (runpy.py:87)
    _run_module_as_main (runpy.py:194)
```

**Comment by arieleiz** [comment]
the code loads a very big parquet dataframe into pandas (with `df = pandas.read_parquet(path)`), after loading the parquet it takes about 5 minutes to finish evaluation, during the entire 5 minutes the stack is as seen above, somehow the T property is invoked which transposes the dataframe, a very expensive operation

**Comment by davidhalter** [comment]
What is the T property? Also what's the code in your terminal when this happens? I have no idea about pandas, I have never used it.

**Comment by arieleiz** [comment]
Pandas is a central python package for machine learning, it stores large "dataframes", which are basically columnar tables that can sometimes be huge. The T property returns a transpose version of the table, which for large tables can take many minutes.
The issue here as far as I understand is that the property is invoked for getting auto-completions, which can block for a long time.

**Comment by PeterJCLaw** [comment]
Having a very quick play with this here's steps to reproduce:

- `pip install pandas, ipython`
- launch `ipython`
- run `import pandas`
- run `df = pandas.DataFrame()`
- type `df.T`, then press tab

This will cause `DirectObjectAccess.getattr_paths` to run in a state where `self._obj` is the DataFrame and call `getattr` on it with `T` as the attribute name. (I determined this by adding a print inline in that method)

I'm not sure what this translates to in terms of the Jedi API calls, though hopefully that translation is relatively easy to do from those steps.

For reference I'm using jedi @ 825c6b93bf67f5b44552357db52a11b3f62cf864, ipython==8.13.2, pandas==2.0.1.

This doesn't reproduce for me under `sith.py`, but I'm guessing that's maybe down to the interpreter use of Jedi which IPython has? (:shrug:)

**Comment by davidhalter** [comment]
Thanks @PeterJCLaw. But in your case you're not loading any data. Is that still slow? It's just a bit surprising, because I would have guessed that this is just happening with bigger datasets.

`sith.py` only uses non-interpreted contexts and therefore does not execute any code (which is why this is slow in the first place).

**Comment by PeterJCLaw** [comment]
Sorry, no, the case of the empty dataset isn't particularly slow. I think the issue here is that jedi is executing user-code at all during the attribute lookup, which for the case of `pandas.DataFrame.T` is potentially a very slow operation. I'm not really sure what Jedi needs to do this lookup for (I've not dug that far into the code), though perhaps fetching the descriptor would be sufficient? That doesn't completely preclude running code, but seems far less likely to actually do so (and would, I think, be a strict subset of the cases where it currently does).

**Comment by davidhalter** [comment]
On master there is `jedi.settings.instance_allow_descriptor_getattr = False`. that should probably fix the issue, right? 

@PeterJCLaw The history of this issue is that we have used the descriptor before, but people complained a lot, so I allowed properties to be accessed again. In general this is a pretty good thing for two reasons:

1. Most people actually like having completions on stuff like `Foo.name.<complete>`
2. Most Python software does not do big calculations in properties, because they look like attributes and a lot of people (like me) think that having compute intensive properties is an anti-pattern.

I think there are two potential fixes for this specific problem without breaking a lot of users behavior:

1. Special case pandas for this one thing.
2. Avoid using getattr when doing completions on the parent object and only execute the descriptor once we're using it on the object. In this concrete example it would mean that the user will not get a slowdown for `pandas.DataFrame().<complete>` and `pandas.DataFrame().T<complete>`, but will get a slowdown for `pandas.DataFrame().T.<complete>`. Note that the latter case is a case where a user at least eventually wants to access the property, so that probably reduces the amount of cases by 90% where we get complaints.

What do you think?

**Comment by PeterJCLaw** [comment]
> On master there is `jedi.settings.instance_allow_descriptor_getattr = False`. that should probably fix the issue, right?

Yeah, from some quick testing (with a class of my own which prints when a property is accessed) that seems to work.

Aside: looking at the usage of that member, in [`CompiledValueFilter._get`](https://github.com/davidhalter/jedi/blob/825c6b93bf67f5b44552357db52a11b3f62cf864/jedi/inference/compiled/value.py#L442-L462) there seems to be an [initial conditional block](https://github.com/davidhalter/jedi/blob/825c6b93bf67f5b44552357db52a11b3f62cf864/jedi/inference/compiled/value.py#L446-L447) which has no functional impact as the conditional block is just `pass`. Is that expected? (Should it be an early `return []`? Can the conditional check & block just be dropped?)

> @PeterJCLaw The history of this issue is that we have used the descriptor before, but people complained a lot, so I allowed properties to be accessed again. In general this is a pretty good thing for multiple:
> 
> 1. Most people actually like having completions on stuff like `Foo.name.<complete>`

Interesting. (I agree completions in that case are a desirable outcome). I'd assumed that the descriptor would be sufficient to know the return type (much as Jedi can work out the return type of a method without actually running the method), though I'm guessing that:

a) there's both a difference here between the static case and the interpreter case (in the latter Jedi is presumably more willing to just run the code and/or less able to introspect things?), and
b) running the code is a fallback where Jedi otherwise cannot see already (e.g: from type annotations) what the return type is going to be?

Certainly I had assumed the latter would be the case, though perhaps it's not. FWIW, the current version of pandas does have type annotations on the `DataFrame` class' definition, including that the `T` property, so I would have assumed that that would be enough to avoid the code execution (both statically and in the interpreter).
If it's intentional that the annotations aren't used it would be interesting to understand why.

> 2. Most Python software does not do big calculations in properties, because they look like attributes and a lot of people (like me) think that having compute intensive properties is an anti-pattern.

Yup, I agree.

> 3. The default Python completer AFAIK also executes stuff like `dataframe.T.` when completing, so people are used to that behavior.

TIL. Seems reasonable, though I wonder how much that was influenced by Jedi setting a precedent? :shrug:

> I think there are two potential fixes for this specific problem without breaking a lot of users behavior:
> 
> 1. Special case pandas for this one thing.
>
> 2. Avoid using getattr when doing completions on the parent object and only execute the descriptor once we're using it on the object. In this concrete example it would mean that the user will not get a slowdown for `pandas.DataFrame().<complete>` and `pandas.DataFrame().T<complete>`, but will get a slowdown for `pandas.DataFrame().T.<complete>`. Note that the latter case is a case where a user at least eventually wants to access the property, so that probably reduces the amount of cases by 90% where we get complaints.

The general solution feels better here. I agree with your reasoning about the slowdown being more reasonable once you're looking at members of the value returned from the property.

However, the slight spanner in the works is that `DataFrame().T` doesn't appear to cache its results. While Jedi does cache things, meaning that subsequent Jedi calls are faster, the user will still pay the cost of creating the transposition again when they run the code "for real". I'm not sure how much this is something we want to care about.

It also looks like Jedi ends up running the property twice in the `DataFrame().T.<complete>` case at the moment (again tested via my printing property rather than via pandas). While perhaps a separate issue, if there's a way to avoid that then doing so would likely improve things for `pandas` users too.

---

I was curious about the relationship with the annotations on the property, so I had a look at the static completions offered after `DataFrame().T.`. It seems that Jedi isn't inferring a type at all for that case. I wonder if that could be related here? That is: if we fixed the static case, would that also mean Jedi wouldn't need to run the member in the interpreter?

**Comment by davidhalter** [comment]
@PeterJCLaw Sorry for the long delay. I'm very busy rewriting Mypy in Rust these days, so I kind of forget about the Jedi issue tracker from time to time.

> Aside: looking at the usage of that member, in CompiledValueFilter._get there seems to be an initial conditional block which has no functional impact as the conditional block is just pass. Is that expected? (Should it be an early return []? Can the conditional check & block just be dropped?)

Thanks for reading the code. This feels really awkward. It's definitely either a bug or just weird code. I will have to dig. I'm definitely not just going to delete it without trying to understand first why it's been there.

> Interesting. (I agree completions in that case are a desirable outcome). I'd assumed that the descriptor would be sufficient to know the return type (much as Jedi can work out the return type of a method without actually running the method), though I'm guessing that:

> a) there's both a difference here between the static case and the interpreter case (in the latter Jedi is presumably more willing to just run the code and/or less able to introspect things?), and
> b) running the code is a fallback where Jedi otherwise cannot see already (e.g: from type annotations) what the return type is going to be?

> Certainly I had assumed the latter would be the case, though perhaps it's not. FWIW, the current version of pandas does have type annotations on the DataFrame class' definition, including that the T property, so I would have assumed that that would be enough to avoid the code execution (both statically and in the interpreter).
If it's intentional that the annotations aren't used it would be interesting to understand why.

It's unfortunately not as easy as you assume. This is something that we should probably fix as well sometimes. I'm not sure how easy it is to do if generics are involved, but we could also try to handle the simple case for now and ignore stuff like forward references, etc. It's just that there's a lot of edge cases here that I'm not sure how to deal with. Consider for example something like `Union[int, T]`. This then either comes as a string (because the user uses `from __future__ import annotations` or a forward string), which is annoying, because we have to parse it and we have to kind of infer the location in our tree with the position of the function (~where we probably get the line number and can then somehow probably figure out the function and from that what types would be valid~ nevermind, we should probably just only search the module scope, which is good enough in 99.9% of cases). But then there's also the other case where we have to parse the `Union[int, T]` and then try to infer the type that way, where T is something defined in an interpreter environment that is an instance of TypeVar (probably or maybe it's something weird like `_SpecialGenericAlias`.

So what I'm trying to say: **It's just a very weird mix of strings and actual Python objects** that is definitely manageable and it's probably the correct way to work with that ASAP, but there's probably a lot of testing and bugs that need to be fixed once we do that.

The current approach is way simpler: We try to evaluate the type until we reach a `()`. So in `a.b.c().d.e`, `a.b.c` are evaluated, then we look for the definition of that, try to execute it statically and look up `d.e` on that statically. With this approach there is no annoying changing between the approaches all the time, it's just evaluation at the start and inference after that with kind of simple rules. However this brings me to the *idea that we could detect this for properties with annotations only*, because everything else works fine with annotations as you mention, it's just that properties have this weird way of being executed implicitly, which we could avoid in the presence of annotations. I'll try to figure something out. 

> > The default Python completer AFAIK also executes stuff like dataframe.T. when completing, so people are used to that behavior.

> TIL. Seems reasonable, though I wonder how much that was influenced by Jedi setting a precedent? 🤷

No it's the other way around. The IPython completions have been there for a long time. IPython uses more or less the completions that something like 

```
     import rlcompleter
    readline.parse_and_bind("tab: complete")
```
is providing in a normal `python3` shell. This has existed for a way longer time than Jedi and was the inspiration for a lot of the early Jedi `Interpreter` completions.

> It also looks like Jedi ends up running the property twice in the DataFrame().T.<complete> case at the moment (again tested via my printing property rather than via pandas). While perhaps a separate issue, if there's a way to avoid that then doing so would likely improve things for pandas users too.

I would have to investigate, it's definitely something that we could look out for, but currently not a high priority for me, but maybe I find this while trying to deal with the general issue here that completions do a getattr on all of its attributes (which is probably not really necessary).

> I was curious about the relationship with the annotations on the property, so I had a look at the static completions offered after DataFrame().T.. It seems that Jedi isn't inferring a type at all for that case. I wonder if that could be related here? That is: if we fixed the static case, would that also mean Jedi wouldn't need to run the member in the interpreter?

~No, this won't help unfortunately. See my explanation a bit before about how generics in Interpreter work.~ I'm so confused, because I work with TypeVars a lot named `T`, so I was assuming `T` is also a TypeVar, but it's really just a simple property https://github.com/pandas-dev/pandas/blob/v2.0.3/pandas/core/frame.py#L3607-L3608.

The reason for why this does not work in the static case might also be that there's a function called `is_big_annoying_library` in Jedi that limits some of the inference to make pandas/numpy and other modules a bit faster. But I'm not sure here and I would have to investigate myself.

-------

I'll definitely try to figure out a new release now and I'm trying to fix a few more things around the REPL. I'm kind of disappointed in myself that I'm not putting in more love there, because this is something that almost everybody at a university in my circles has used (through Jupyter Notebook AFAIK). It's kind of shocking how many people use it, while there's almost no tests in Jedi for that feature (most tests are about static analysis). It's probably also the reason why Jedi is slowly approaching 1 billion downloads.

So I'll try to set aside a bit more time for that.

**Comment by davidhalter** [comment]
> It also looks like Jedi ends up running the property twice in the DataFrame().T.<complete> case at the moment (again tested via my printing property rather than via pandas). While perhaps a separate issue, if there's a way to avoid that then doing so would likely improve things for pandas users too.

I just realized that this is probably expected and unavoidable, because `DataFrame().T` returns a `DataFrame()` again. Completions on it just cause us to access `DataFrame().T.T`.

**Comment by davidhalter** [comment]
> > It also looks like Jedi ends up running the property twice in the DataFrame().T.<complete> case at the moment (again tested via my printing property rather than via pandas). While perhaps a separate issue, if there's a way to avoid that then doing so would likely improve things for pandas users too.

> I just realized that this is probably expected and unavoidable, because `DataFrame().T` returns a `DataFrame()` again. Completions on it just cause us to access `DataFrame().T.T`.

I realized, that I'm wrong, sorry for the confusion. In the case of completing `DataFrame().T.T`, there are indeed three accesses instead of two. I don't feel like this is really needed, but it's a bit too much work to build it differently for me to quickly to it, so I'm leaving it for now. I guess most computationally intensive property accesses will be behind caches as well (like Pandas).

**Comment by davidhalter** [comment]
All right. I feel like I finally fixed all the issues. It was quite a few commits, but the 4 relevant ones are linked above. This will improve pretty much all cases that we struggled with previously and we will probably have far fewer people complaining about performance.

@PeterJCLaw I feel like this also fixes all the issues and implements the ideas you brought up. Thanks for challenging me.

**Comment by davidhalter** [comment]
I just released v0.19.0. Feel free to upgrade and let me know if there are further issues.
